java/java.hints/src/org/netbeans/modules/java/hints/suggestions/Tiny.java

```diff
@@ -400,4 +402,18 @@ protected void performRewrite(final TransformationContext ctx) {
 
     }
 
+    @Hint(displayName = "#DN_org.netbeans.modules.java.hints.suggestions.Tiny.inlineRedundantVar", description = "#DESC_org.netbeans.modules.java.hints.suggestions.Tiny.inlineRedundantVar", category="suggestions", severity=Severity.HINT)
+    @TriggerPattern(value="$before$; $type $var = $init; $statements$; return $var;")
+    public static ErrorDescription inlineRedundantVar(HintContext ctx) {
+        TreePath var = ctx.getVariables().get("$var");
+        Collection<? extends TreePath> stats = ctx.getMultiVariables().get("$statements$");
+        if (var != null && stats != null) {
+            Element element = ctx.getInfo().getTrees().getElement(var);
+            if (element != null && element.getKind() == ElementKind.LOCAL_VARIABLE && !Utilities.isReferencedIn(ctx.getInfo(), var, stats)) {
```

Comment:
Q: what if `$statements` have some effects on the data used in the  `$init` expression ?

Reply:
Good point. Hint changed to a more restrictive version - limited only to
- local variables that are immediately returned
- local variables that are immediately assigned to another variable and then not used

Reply:
could this be solved by listing all local vars of $init (maybe there is a utility for that) and running `!referencedIn` on $statements$ for each of them?

Reply:
@mbien  I believe it can't be computed in general. If `$statements$` have some non-local side effect, then even a method call from `$init` may be affected through some field of another object.

There are at least two solutions:
- simple one: do not call it `redundant` variable, but simply suggest to inline the expression :)) It's the user responsibility to decide, and IDE to do the refactoring
- make the constraints tight:
-- does not reference any potentially non-final symbols nonlocal (that includes any invoked methods), or `statements` does not make any outbound method/constructor calls. We do not have machinery for nonlocal flow analysis.
-- `statements` does not *assign* to *any* of the symbols used in `init`

Reply:
I like your first solution. I maintain a list of jackpot rules I used myself over time (https://github.com/mbien/jackpot-inspections) and the first FAQ entry is:

Q: Can I apply suggested inspections without code review and testing?
A: No.

I don't think all hints have to be fool proof it it is communicated to the user that they can be risky. Like switching out StringBuffer with Builder will work in 99% of the time, but fail in the one case when someone actually shared it between threads. I believe this is completely OK.

Thats why I am somewhat for reverting it to the more aggressive version which uses `!referencedIn` as best effort attempt to suggest the right thing.